drivers/spi/spi_dw.h

```diff
@@ -35,52 +49,125 @@ struct spi_dw_data {
 	uint16_t _unused;
 };
 
+/* Register operation funcitons */
+static uint32_t aux_reg_read(uint8_t size, uint32_t addr, uint32_t off)
+{
+	switch (size) {
+	case 8:
+		return sys_in8(addr + off/4);
+	case 16:
+		return sys_in16(addr + off/4);
+	case 32:
+		return sys_in32(addr + off/4);
+	default:
+		return -EINVAL;
+	}
+}
+
+static void aux_reg_write(uint8_t size, uint32_t data, uint32_t addr, uint32_t off)
+{
+	switch (size) {
+	case 8:
+		sys_out8(data, addr + off/4); break;
+	case 16:
+		sys_out16(data, addr + off/4); break;
+	case 32:
+		sys_out32(data, addr + off/4); break;
+	default:
+		break;
+	}
+}
+
+static void aux_reg_set_bit(uint8_t bit, uint32_t addr, uint32_t off)
+{
+	sys_io_set_bit(addr + off/4, bit);
+}
+
+static void aux_reg_clear_bit(uint8_t bit, uint32_t addr, uint32_t off)
+{
+	sys_io_clear_bit(addr + off/4, bit);
+}
+
+static int aux_reg_test_bit(uint8_t bit, uint32_t addr, uint32_t off)
+{
+	return sys_io_test_bit(addr + off/4, bit);
+}
+
+static uint32_t reg_read(uint8_t size, uint32_t addr, uint32_t off)
+{
+	switch (size) {
+	case 8:
+		return sys_read8(addr + off);
+	case 16:
+		return sys_read16(addr + off);
+	case 32:
+		return sys_read32(addr + off);
+	default:
+		return -EINVAL;
+	}
+}
+
+static void reg_write(uint8_t size, uint32_t data, uint32_t addr, uint32_t off)
+{
+	switch (size) {
+	case 8:
+		sys_write8(data, addr + off); break;
+	case 16:
+		sys_write16(data, addr + off); break;
+	case 32:
+		sys_write32(data, addr + off); break;
+	default:
+		break;
+	}
+}
+
+static void reg_set_bit(uint8_t bit, uint32_t addr, uint32_t off)
+{
+	sys_set_bit(addr + off, bit);
+}
+
+static void reg_clear_bit(uint8_t bit, uint32_t addr, uint32_t off)
+{
+	sys_clear_bit(addr + off, bit);
+}
+
+static int reg_test_bit(uint8_t bit, uint32_t addr, uint32_t off)
+{
+	return sys_test_bit(addr + off, bit);
+}
+
 /* Helper macros */
 
 #define SPI_DW_CLK_DIVIDER(clock_freq, ssi_clk_hz) \
 		((clock_freq / ssi_clk_hz) & 0xFFFF)
 
-#ifdef CONFIG_SPI_DW_ARC_AUX_REGS
-#define Z_REG_READ(__sz) sys_in##__sz
-#define Z_REG_WRITE(__sz) sys_out##__sz
-#define Z_REG_SET_BIT sys_io_set_bit
-#define Z_REG_CLEAR_BIT sys_io_clear_bit
-#define Z_REG_TEST_BIT sys_io_test_bit
-#else
-#define Z_REG_READ(__sz) sys_read##__sz
-#define Z_REG_WRITE(__sz) sys_write##__sz
-#define Z_REG_SET_BIT sys_set_bit
-#define Z_REG_CLEAR_BIT sys_clear_bit
-#define Z_REG_TEST_BIT sys_test_bit
-#endif /* CONFIG_SPI_DW_ARC_AUX_REGS */
-
 #define DEFINE_MM_REG_READ(__reg, __off, __sz)				\
-	static inline uint32_t read_##__reg(uint32_t addr)			\
+	static inline uint32_t read_##__reg(spi_dw_read_t read_func, uint32_t addr)	\
```

Comment:
since the function is always info->read_func and since addr is always info->regs, let's switch to read_##__reg(struct spi_dw_config *info)
It will be cleaner in the driver side (it will keep the function calls short).

Apply the same on all macros below